snmp_types.yaml

```diff
@@ -0,0 +1,87 @@
+tosca_definitions_version: cloudify_dsl_1_0
+
+imports:
+  - http://www.getcloudify.org/spec/cloudify/3.2m4/types.yaml
+  - http://www.getcloudify.org/spec/diamond-plugin/1.2m4/plugin.yaml
+  - http://www.getcloudify.org/spec/openstack-plugin/1.2m4/plugin.yaml
+
+
+node_types:
+  SNMPSecurityGroup:
+    derived_from: cloudify.openstack.nodes.SecurityGroup
+    properties:
+      security_group:
+        default:
+          name: snmp_security_group
+          description: Allows UDP communication for SNMP
+      rules:
+        default:
+          - remote_ip_prefix: 0.0.0.0/0
+            port: 161
+            protocol: udp
+
+  SNMPProxy:
+    derived_from: cloudify.nodes.Root
+    interfaces:
+      cloudify.interfaces.lifecycle:
+        create:
+          implementation: scripts/install_requirements.sh
+      cloudify.interfaces.monitoring:
+        start:
+          implementation: diamond.diamond_agent.tasks.add_collectors
+          inputs:
+            collectors_config:
+              default:
+                SNMPProxyCollector:
+                  path: collectors/snmpproxy.py
+                  config: { get_attribute: [SELF, snmp_collector_config] }
+
+  SnmpdConfiguringNode:
+    derived_from: cloudify.nodes.Root
+    interfaces:
+      cloudify.interfaces.lifecycle:
+        configure: scripts/setup_snmpd.sh
+
+  SNMPManagerProxy:
+    derived_from: cloudify.nodes.Compute # "fake compute"
+    properties:
+      install_agent:
+        default: false
+      diamond_config:
+        default:
+          interval: 1
+    interfaces:
+      cloudify.interfaces.lifecycle:
+        create:
+          executor: central_deployment_agent
+          implementation: diamond.diamond_agent.tasks.install
+          inputs:
+            diamond_config:
+              default: { get_property: [SELF, diamond_config] }
+        configure:
+          executor: central_deployment_agent
+          implementation: scripts/install_requirements.sh
+        start:
+          executor: central_deployment_agent
+          implementation: diamond.diamond_agent.tasks.start
+        stop:
+          executor: central_deployment_agent
+          implementation: diamond.diamond_agent.tasks.stop
+        delete:
+          executor: central_deployment_agent
+          implementation: scripts/diamond_clean_up.py
+      cloudify.interfaces.monitoring:
+        start:
+          implementation: diamond.diamond_agent.tasks.add_collectors
+          executor: central_deployment_agent
+          inputs:
+            collectors_config:
+              default:
+                SNMPProxyCollector:
+                  path: collectors/snmpproxy.py
+                  config: { get_attribute: [SELF, snmp_collector_config ] }
+
+relationships:
```

Comment:
You can extend the relationship definition:
```
relationships:
  cloudify.relationships.monitors:
    derived_from: cloudify.relationships.depends_on
    source_interfaces:
      cloudify.interfaces.relationship_lifecycle:
        preconfigure:
          implementation: scripts/append_diamond_conf.py
          executor: central_deployment_agent
          inputs:
            port:
              default: 161
              description: snmp port
            community:
              description: the community
            path_suffix:
              description: the path suffix ?
            oids:
              description: the oids to use
```

This way, node templates that use this relationship only need to specify the operation inputs, and only those that that don't have a default value. In addition, defining the operation schema like we did here also provides validation, so that if a user did not configure one of the required inputs (those without default), the blueprint validation will fail during the blueprint upload phase

A usage example could then be:
```
  ProxyNode:
    type: SNMPProxy
    relationships:
      - type: cloudify.relationships.monitors
        target: SNMPMonitoredHost
        source_interfaces:
          cloudify.interfaces.relationship_lifecycle:
            preconfigure:
              inputs:
                community: public
                path_suffix: cpu
                oids:
                  1.3.6.1.4.1.2021.10.1.3.1: total.system
                  1.3.6.1.4.1.2021.10.1.3.2: total.user
```

The documentation should be updated as well to reflect this changes